[cutlass backend] changes how gpu_kernels_o are handled for cutlass

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155875

Currently, we do it a bit hacky: Look at all the .o we have from this session, add them all to AOTI. This for example doesn't work if we do multiple AOTI compilation in one session, without clearing the inductor cache.

Also I want to change how cutlass .so are compiled. Hence this change.

This change is broken down since @coconutruben is trying to make a change to the same files too.

Differential Revision: [D76563003](https://our.internmc.facebook.com/intern/diff/D76563003/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] changes how gpu_kernels_o are handled for cutlass



Differential Revision: [D76563003](https://our.internmc.facebook.com/intern/diff/D76563003/)

[ghstack-poisoned]

ColinPeppler commented on torch/_inductor/codecache.py:
```diff
@@ -2080,14 +2080,18 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
-                if entry.output_path.endswith(".o")
-            ]
+            gpu_kernels_o = []
+            if torch.version.hip:
+                gpu_kernels_o.extend(
+                    entry.output_path
+                    for entry in ROCmCodeCache.cache.values()
+                    if entry.output_path.endswith(".o")
+                )
+            else:
+                gpu_kernels_o.extend(CUDACodeCache.aot_kernels_o)
+                # clear the list of aot kernels after each linking
+                CUDACodeCache.aot_kernels_o.clear()
```
can we just add `aot_kernels_o` to `ROCmCodeCache` so we can avoid this branching?

henrylhtsang committed:
Update on "[cutlass backend] changes how gpu_kernels_o are handled for cutlass"


Currently, we do it a bit hacky: Look at all the .o we have from this session, add them all to AOTI. This for example doesn't work if we do multiple AOTI compilation in one session, without clearing the inductor cache.

Also I want to change how cutlass .so are compiled. Hence this change.

This change is broken down since coconutruben is trying to make a change to the same files too.

Differential Revision: [D76563003](https://our.internmc.facebook.com/intern/diff/D76563003/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
